grakn-graph/src/main/java/ai/grakn/graph/internal/RelationEdge.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Grakn - A Distributed Semantic Database
+ * Copyright (C) 2016  Grakn Labs Limited
+ *
+ * Grakn is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Grakn is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Grakn. If not, see <http://www.gnu.org/licenses/gpl.txt>.
+ */
+
+package ai.grakn.graph.internal;
+
+import ai.grakn.concept.ConceptId;
+import ai.grakn.concept.LabelId;
+import ai.grakn.concept.Relation;
+import ai.grakn.concept.RelationType;
+import ai.grakn.concept.Role;
+import ai.grakn.concept.Thing;
+import ai.grakn.util.Schema;
+import com.google.common.collect.Sets;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * <p>
+ *     Encapsulates The {@link Relation} as a {@link EdgeElement}
+ * </p>
+ *
+ * <p>
+ *     This wraps up a {@link Relation} as a {@link EdgeElement}. It is used to represent any binary {@link Relation}.
+ *     This also includes the ability to automatically reify a {@link RelationEdge} into a {@link RelationReified}.
+ * </p>
+ *
+ * @author fppt
+ *
+ */
+class RelationEdge implements RelationStructure{
+    private final EdgeElement edgeElement;
+
+    private final Cache<RelationType> relationType = new Cache<>(() ->
+            edge().graph().getOntologyConcept(LabelId.of(edge().property(Schema.EdgeProperty.RELATION_TYPE_LABEL_ID))));
+
+    private final Cache<Role> ownerRole = new Cache<>(() -> edge().graph().getOntologyConcept(LabelId.of(
+            edge().property(Schema.EdgeProperty.RELATION_ROLE_OWNER_LABEL_ID))));
+
+    private final Cache<Role> valueRole = new Cache<>(() -> edge().graph().getOntologyConcept(LabelId.of(
+            edge().property(Schema.EdgeProperty.RELATION_ROLE_VALUE_LABEL_ID))));
+
+    private final Cache<Thing> owner = new Cache<>(() -> edge().graph().factory().buildConcept(edge().source()));
+    private final Cache<Thing> value = new Cache<>(() -> edge().graph().factory().buildConcept(edge().target()));
+
+    RelationEdge(EdgeElement edgeElement) {
+        this.edgeElement = edgeElement;
+    }
+
+    RelationEdge(RelationType relationType, Role ownerRole, Role valueRole, EdgeElement edgeElement) {
+        this(edgeElement);
+
+        edgeElement.propertyImmutable(Schema.EdgeProperty.RELATION_ROLE_OWNER_LABEL_ID, ownerRole, null, o -> o.getLabelId().getValue());
```

Comment:
Possible to get rid of `RELATION_ROLE_OWNER_LABEL_ID` and `RELATION_ROLE_VALUE_LABEL_ID` as this information is already stored in it's type. For another PR though.